### PR TITLE
Improve documentation for options

### PR DESCRIPTION
**Pull request by konsumlamm**
* improve wording
* clarify behaviour for pointers
* improve `runnableExamples`

### Future work

* [x] link [`fusion/matching`](https://nim-lang.github.io/fusion/src/fusion/matching.html) and add examples for pattern matching on `Option`s (#17120)

### CODE REVIEW

File: lib/pure/options.nim [hunk 1]
```diff
@@ -338,15 +342,14 @@ proc `==`*(a, b: Option): bool {.inline.} =
     assert b == d
     assert not (a == b)
 
-  (a.isSome and b.isSome and a.val == b.val) or (not a.isSome and not b.isSome)
+  (a.isSome and b.isSome and a.val == b.val) or (a.isNone and b.isNone)
```
Comment (timotheecour) [inline lib/pure/options.nim]: optimization:
```suggestion
  when a.val is SomePointer: a.val == b.val
  else: (a.isSome and b.isSome and a.val == b.val) or (a.isNone and b.isNone)
```

File: lib/pure/options.nim [hunk 2]
```diff
@@ -286,9 +291,9 @@ proc flatMap*[A, B](self: Option[A],
   runnableExamples:
     proc doublePositives(x: int): Option[int] =
       if x > 0:
-        return some(2*x)
+        some(2 * x)
       else:
-        return none(int)
+        none(int)
     let
```
Comment (timotheecour) [inline lib/pure/options.nim]: https://github.com/timotheecour/Nim/issues/534#issuecomment-780791150

File: lib/pure/options.nim [hunk 3]
```diff
@@ -234,49 +236,52 @@ proc map*[T, R](self: Option[T], callback: proc (input: T): R): Option[R] {.inli
...
-proc flatten*[A](self: Option[Option[A]]): Option[A] {.inline.} =
+proc flatten*[T](self: Option[Option[T]]): Option[T] {.inline.} =
   ## Remove one level of structure in a nested `Option`.
+  ##
+  ## **See also:**
+  ## * `flatMap proc <#flatMap,Option[T],proc(T)>`_
   runnableExamples:
-    let a = some(some(42))
-    assert $flatten(a) == "Some(42)"
+    let
```
Comment (timotheecour) [inline lib/pure/options.nim]: https://github.com/timotheecour/Nim/issues/534#issuecomment-780791150

File: lib/pure/options.nim [hunk 4]
```diff
@@ -234,49 +236,52 @@ proc map*[T, R](self: Option[T], callback: proc (input: T): R): Option[R] {.inli
...
       a = some(42)
       b = none(int)
 
     proc isEven(x: int): bool =
       x mod 2 == 0
 
-    assert $(a.map(isEven)) == "Some(true)"
-    assert $(b.map(isEven)) == "None[bool]"
+    assert a.map(isEven) == some(true)
```
Comment (timotheecour) [inline lib/pure/options.nim]: https://github.com/timotheecour/Nim/issues/534#issuecomment-780791150

plus wherever relevant

File: lib/pure/options.nim [hunk 5]
```diff
@@ -338,15 +310,17 @@ proc `==`*(a, b: Option): bool {.inline.} =
     assert b == d
     assert not (a == b)
 
-  (a.isSome and b.isSome and a.val == b.val) or (not a.isSome and not b.isSome)
+  when T is SomePointer:
+    a.val == b.val
+  else:
+    (a.isSome and b.isSome and a.val == b.val) or (a.isNone and b.isNone)
```
Comment (dom96) [inline lib/pure/options.nim:316]: This isn't just a doc change, maybe it should go into a separate PR?

**Review by dom96 — APPROVED** [review]
Just one remark, the rest looks good.

File: lib/pure/options.nim (hunk 5 above)
Comment (konsumlamm) [reply lib/pure/options.nim:316]: Hmm, that was suggested by @timotheecour, however I don't think such a small change warrants a new PR. What if I just change the PR description?

Comment (timotheecour) [reply lib/pure/options.nim:316]: i had suggested that change; I think it's simple enough to be grouped in same PR but I don't mind either way

File: lib/pure/options.nim [hunk 6]
```diff
@@ -338,15 +310,17 @@ proc `==`*(a, b: Option): bool {.inline.} =
...
 
 proc `$`*[T](self: Option[T]): string =
   ## Get the string representation of the `Option`.
-  ##
-  ## If the `Option` has a value, the result will be `Some(x)` where `x`
-  ## is the string representation of the contained value.
-  ## If the `Option` does not have a value, the result will be `None[T]`
-  ## where `T` is the name of the type contained in the `Option`.
+  runnableExamples:
+    assert $some(42) == "Some(42)"
```
Comment (timotheecour) [inline lib/pure/options.nim:321]: unrelated to this PR but in future work IMO this should change to `some(42)`, `none(int)` for consistency with `none`, `some` (None, Some don't exist and could potentially be other types)

EDIT => PR https://github.com/nim-lang/Nim/pull/17147

**Review by timotheecour — APPROVED** [review]

**Comment by konsumlamm** [comment]
@timotheecour I think you can remove the `TODO: followup needed` label now (unless you were referring to https://github.com/nim-lang/Nim/pull/17036#discussion_r577988573).

**Comment by timotheecour** [comment]
I was referring to https://github.com/nim-lang/Nim/pull/17036#discussion_r577988573 indeed, but here's the PR: https://github.com/nim-lang/Nim/pull/17147
